File: lib/StateUI.Maui/Sources/Rendering/MotionLaw.cs
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

using StateUI.Maui.Protocol;

namespace StateUI.Maui.Rendering;

/// <summary>
/// Where a value is at a given moment of its motion, and how fast it is going
/// there - and the curves an eased motion follows.
/// </summary>
/// <remarks>
/// <para>
/// Every law here is CLOSED FORM in the time since the motion began: nothing is
/// integrated frame by frame, so a run of frames answers the same numbers
/// whatever the frames were - which is what makes a trajectory a pure function
/// of <c>t</c>, testable against a hand-written clock, and unharmed by a
/// suspended application (it slews to the end rather than resuming mid-air).
/// </para>
/// <para>
/// Velocity is the law's own derivative, never a difference between two
/// samples. That is what a retarget needs: the motion that replaces this one
/// starts from the speed this one had, so the value bends instead of being cut.
/// A curve's SLOPE is a central difference on the CURVE - a pure function of
/// how far through the motion is - and not a difference between two frames, so
/// it answers the same number in every run and every process, at any frame
/// rate, including none at all.
/// </para>
/// <para>
/// The laws and their curves are the core's: a line-by-line copy of the Swift
/// <c>HostMotionLaw</c>, in the same arithmetic order, which every runtime
/// walks with - a value walked on <c>cubicOut</c> here is the sequence of
/// numbers every other runtime walks. <c>motion-laws.txt</c>, the core's
/// trajectory table, is what proves the copy (<c>MotionLawTests</c>).
/// </para>
/// </remarks>
internal static class MotionLaw
{
    /// <summary>
    /// A spring is at rest when it is this close to its target and this slow -
    /// in the value's own units, and per millisecond for the speed.
    /// </summary>
    /// <remarks>
    /// One number for every lane, which is why lanes are kept in the units a
    /// reader sees: a colour channel is 0-1, a coordinate is a unit on screen.
    /// A thousandth of either is under any screen's resolution.
    /// </remarks>
    internal const double Still = 0.001;

    /// <summary>
    /// However slow it gets, a spring - the one law with no stated end - is
    /// over after this many milliseconds, so nothing can hold the frame clock
    /// awake for ever.
    /// </summary>
    internal const double Longest = 10_000;

    /// <summary>Whether two runs of lanes stand at the same place, to <see cref="Still"/>.</summary>
    /// <param name="left">One run.</param>
    /// <param name="right">The other, as long.</param>
    /// <returns>Whether every lane agrees.</returns>
    internal static bool Same(double[] left, double[] right)
    {
        for (int lane = 0; lane < left.Length; lane++)
        {
            if (Math.Abs(left[lane] - right[lane]) >= Still)
            {
                return false;
            }
        }

        return true;
    }

    /// <summary>
    /// Puts the value and its speed at <paramref name="t"/> into
    /// <paramref name="p"/> and <paramref name="v"/>.
    /// </summary>
    /// <remarks>
    /// The motion's numbers are read as the wire carries them, a spring's
    /// response held to a millisecond and its damping off nought here, as the
    /// core's law holds them.
    /// </remarks>
    /// <param name="trip">The trip, holding where it started, where it is going and how.</param>
    /// <param name="t">Milliseconds since the motion began.</param>
    /// <param name="p">Filled with the value at that moment.</param>
    /// <param name="v">Filled with how fast each lane is moving, per millisecond.</param>
    /// <returns>Whether the motion is over.</returns>
    internal static bool Sample(Trip trip, double t, double[] p, double[] v)
    {
        HostMotion motion = trip.Motion;

        return motion.Kind switch
        {
            HostMotion.Law.Spring => Spring(
                trip, Math.Max(motion.Millis, 1u), Math.Max(motion.Factor, 0.01), t, p, v),
            _ => Eased(trip, motion.Millis, motion.Curve, t, p, v),
        };
    }

    /// <summary>How far along a curve is at <paramref name="s"/>, from 0 to 1.</summary>
    /// <remarks>
    /// A number the wire does not name is the straight line.
    /// </remarks>
    /// <param name="curve">The curve.</param>
    /// <param name="s">How far through the motion is, from 0 to 1.</param>
    /// <returns>The fraction of the distance covered - which the bouncing and
    /// springing curves deliberately take past 1 and back.</returns>
    internal static double Ease(HostEasing curve, double s)
    {
        double x = Math.Clamp(s, 0, 1);

        switch (curve)
        {
            case HostEasing.SineOut:
                return Math.Sin(x * Math.PI / 2);

            case HostEasing.SineIn:
                return 1 - Math.Cos(x * Math.PI / 2);

            case HostEasing.SineInOut:
                return (1 - Math.Cos(x * Math.PI)) / 2;

            case HostEasing.CubicIn:
                return x * x * x;

            case HostEasing.CubicOut:
            {
                double shifted = x - 1;
                return (shifted * shifted * shifted) + 1;
            }

            case HostEasing.CubicInOut:
            {
                if (x < 0.5)
                {
                    return 4 * x * x * x;
                }

                double shifted = (2 * x) - 2;
                return (shifted * shifted * shifted / 2) + 1;
            }

            case HostEasing.BounceOut:
                return BounceOut(x);

            case HostEasing.BounceIn:
                return 1 - BounceOut(1 - x);

            case HostEasing.SpringIn:
                return x * x * ((2.70158 * x) - 1.70158);

            case HostEasing.SpringOut:
            {
                double shifted = x - 1;
                return (shifted * shifted * ((2.70158 * shifted) + 1.70158)) + 1;
            }

            default:
                return x;
        }
    }

    /// <summary>How steeply the curve is rising at <paramref name="s"/>.</summary>
    /// <remarks>
    /// Per unit of <c>s</c>, so a caller divides by the motion's length to get
    /// a speed. The step is small enough to be exact for every curve here and
    /// wide enough that no curve's own arithmetic shows through it; at the ends
    /// the difference is one-sided, since there is no curve outside 0 to 1.
    /// </remarks>
    /// <param name="curve">The curve.</param>
    /// <param name="s">How far through the motion is, from 0 to 1.</param>
    /// <returns>The curve's rate of climb there.</returns>
    internal static double Slope(HostEasing curve, double s)
    {
        const double step = 1e-4;

        double from = Math.Clamp(s - step, 0, 1);
        double to = Math.Clamp(s + step, 0, 1);

        if (to - from <= 0)
        {
            return 0;
        }

        return (Ease(curve, to) - Ease(curve, from)) / (to - from);
    }

    /// <summary>
    /// A stated length on a stated curve - or, where the motion it replaced
    /// left speed behind, the cubic that carries that speed into it.
    /// </summary>
    /// <remarks>
    /// FROM REST it is exactly the curve the author asked for, evaluated as
    /// the core evaluates it. With speed at the start it is a Hermite:
    /// the same duration, beginning at the value and the speed the previous
    /// motion had reached, ending at the target at a standstill. So a target
    /// changed mid-walk bends the motion rather than cutting it, and a motion
    /// that nothing interrupted is unchanged.
    /// </remarks>
    private static bool Eased(
        Trip trip, double length, HostEasing curve, double t, double[] p, double[] v)
    {
        if (length <= 0 || t >= length)
        {
            trip.Target.CopyTo(p, 0);
            Array.Clear(v);
            return true;
        }

        double s = t / length;
        double eased = Ease(curve, s);
        double slope = Slope(curve, s);

        // The Hermite basis, which only the lanes that carry speed need.
        double h00 = ((2 * s) - 3) * s * s + 1;
        double h10 = ((s - 2) * s + 1) * s;
        double h01 = (3 - (2 * s)) * s * s;
        double d00 = (6 * s * s) - (6 * s);
        double d10 = (3 * s * s) - (4 * s) + 1;
        double d01 = (6 * s) - (6 * s * s);

        for (int lane = 0; lane < p.Length; lane++)
        {
            double from = trip.From[lane];
            double to = trip.Target[lane];
            double speed = trip.StartV[lane];

            if (speed == 0)
            {
                p[lane] = from + ((to - from) * eased);
                v[lane] = (to - from) * slope / length;
                continue;
            }

            p[lane] = (h00 * from) + (h10 * length * speed) + (h01 * to);
            v[lane] = ((d00 * from) + (d10 * length * speed) + (d01 * to)) / length;
        }

        return false;
    }

    /// <summary>
    /// A mass on a spring, in closed form - critically damped unless the author
    /// bought the overshoot.
    /// </summary>
    /// <remarks>
    /// Written about the DISTANCE LEFT rather than the value, which is what
    /// makes the three damping cases the textbook ones and keeps the target out
    /// of the exponentials.
    /// </remarks>
    private static bool Spring(
        Trip trip, double response, double damping, double t, double[] p, double[] v)
    {
        double w = 2 * Math.PI / response;
        double zeta = damping;
        bool rested = t >= Longest;

        for (int lane = 0; lane < p.Length; lane++)
        {
            double to = trip.Target[lane];
            double x0 = trip.From[lane] - to;
            double v0 = trip.StartV[lane];
            double x, dx;

            if (Math.Abs(zeta - 1) < 1e-6)
            {
                double b = v0 + (w * x0);
                double decay = Math.Exp(-w * t);

                x = (x0 + (b * t)) * decay;
                dx = (b - (w * (x0 + (b * t)))) * decay;
            }
            else if (zeta < 1)
            {
                double wd = w * Math.Sqrt(1 - (zeta * zeta));
                double a = x0;
                double b = (v0 + (zeta * w * x0)) / wd;
                double decay = Math.Exp(-zeta * w * t);
                double cos = Math.Cos(wd * t);
                double sin = Math.Sin(wd * t);

                x = decay * ((a * cos) + (b * sin));
                dx = decay * ((-zeta * w * ((a * cos) + (b * sin)))
                    + (wd * ((b * cos) - (a * sin))));
            }
            else
            {
                double root = w * Math.Sqrt((zeta * zeta) - 1);
                double r1 = -(w * zeta) + root;
                double r2 = -(w * zeta) - root;
                double c1 = (v0 - (r2 * x0)) / (r1 - r2);
                double c2 = x0 - c1;

                x = (c1 * Math.Exp(r1 * t)) + (c2 * Math.Exp(r2 * t));
                dx = (c1 * r1 * Math.Exp(r1 * t)) + (c2 * r2 * Math.Exp(r2 * t));
            }

            p[lane] = to + x;
            v[lane] = dx;

            if (Math.Abs(x) > Still || Math.Abs(dx) > Still)
            {
                continue;
            }

            // Near enough to be over, and every lane has to agree before the
            // whole trip is: a spring on four lanes settles them one by one.
            p[lane] = to;
            v[lane] = 0;
        }

        if (!rested)
        {
            rested = true;

            for (int lane = 0; lane < p.Length && rested; lane++)
            {
                rested = v[lane] == 0 && p[lane] == trip.Target[lane];
            }
        }

        if (rested)
        {
            trip.Target.CopyTo(p, 0);
            Array.Clear(v);
        }

        return rested;
    }

    /// <summary>The bounce that settles into its end, four arcs smaller each time.</summary>
    private static double BounceOut(double x)
    {
        if (x < 1 / 2.75)
        {
            return 7.5625 * x * x;
        }

        if (x < 2 / 2.75)
        {
            double shifted = x - (1.5 / 2.75);
            return (7.5625 * shifted * shifted) + 0.75;
        }

        if (x < 2.5 / 2.75)
        {
            double shifted = x - (2.25 / 2.75);
            return (7.5625 * shifted * shifted) + 0.9375;
        }

        double last = x - (2.625 / 2.75);
        return (7.5625 * last * last) + 0.984375;
    }
}

File: lib/StateUI/Sources/Core/Journey/MotionLaw.swift
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

// The platform's C maths library, for the spring's `exp`, `sin` and `cos`.
#if canImport(Darwin)
import Darwin
#elseif canImport(Android)
import Android
#elseif canImport(Glibc)
import Glibc
#elseif canImport(CRT)
import CRT
#endif

// The two motion laws as numbers, in closed form in the time since an animation
// began. Every runtime animates with these; `Tests/Fixtures/motion-laws.txt`
// pins their numbers.
// Design: docs/design/core/journeys.md#motion-laws

/// Where an animation stands at one instant of its law.
@_spi(Host) public struct HostMotionSample: Equatable, Sendable {
    /// The value of each lane.
    public let value: [Double]

    /// How fast each lane is going, per millisecond.
    public let velocity: [Double]

    /// Whether the animation has arrived: every lane at its destination and still.
    public let rested: Bool
}

/// The two motion laws, evaluated from the time since an animation began.
///
/// Time is in milliseconds, the unit of `Motion`'s numbers, so velocity is per
/// millisecond. A journey reports its velocity per second; an animator converts
/// where it reports.
@_spi(Host) public enum HostMotionLaw {
    /// How near its destination and how slow a lane is when it counts as arrived: in
    /// the value's own units, and per millisecond for the speed. A thousandth of a
    /// colour channel or of a point is below any display's resolution.
    public static let still = 0.001

    /// The longest a spring animates, in milliseconds. A spring has no stated end;
    /// this keeps any animation from holding a display clock awake for ever.
    public static let longest = 10_000.0

    /// Where an animation stands `elapsed` milliseconds after it began.
    ///
    /// An eased animation from a standstill follows its curve exactly. A lane that
    /// began moving follows the cubic Hermite of the same length instead, from the
    /// value and speed it had to its destination at a standstill. A spring settles
    /// each lane on its own, and the animation rests when all of them have.
    ///
    /// - Parameters:
    ///   - motion: The law and its numbers.
    ///   - elapsed: Milliseconds since the animation began.
    ///   - from: Where each lane began.
    ///   - destination: Where each lane is going.
    ///   - velocity: How fast each lane was going when it began, per
    ///     millisecond.
    /// - Returns: The value and velocity at that instant, and whether the
    ///   animation has arrived. Lanes that do not pair up land at once.
    public static func sample(
        _ motion: Motion,
        elapsed: Double,
        from: [Double],
        destination: [Double],
        velocity: [Double]
    ) -> HostMotionSample {
        guard from.count == destination.count, from.count == velocity.count else {
            return landed(destination)
        }

        switch motion.law {
        case .eased:
            return eased(
                length: Double(motion.millis),
                curve: motion.curve,
                elapsed: elapsed,
                from: from,
                destination: destination,
                velocity: velocity)
        case .spring:
            return spring(
                response: Double(max(motion.millis, 1)),
                damping: max(motion.factor, 0.01),
                elapsed: elapsed,
                from: from,
                destination: destination,
                velocity: velocity)
        }
    }

    /// An animation at its destination, standing still.
    private static func landed(_ destination: [Double]) -> HostMotionSample {
        HostMotionSample(
            value: destination,
            velocity: Array(repeating: 0, count: destination.count),
            rested: true)
    }

    private static func eased(
        length: Double,
        curve: Easing,
        elapsed: Double,
        from: [Double],
        destination: [Double],
        velocity: [Double]
    ) -> HostMotionSample {
        guard length > 0, elapsed < length else { return landed(destination) }

        let progress = elapsed / length
        let amount = ease(curve, at: progress)
        let climb = slope(curve, at: progress)

        // The Hermite basis and its derivative, which only a lane that began
        // moving needs.
        let h00 = ((2 * progress) - 3) * progress * progress + 1
        let h10 = ((progress - 2) * progress + 1) * progress
        let h01 = (3 - (2 * progress)) * progress * progress
        let d00 = (6 * progress * progress) - (6 * progress)
        let d10 = (3 * progress * progress) - (4 * progress) + 1
        let d01 = (6 * progress) - (6 * progress * progress)

        var value = Array(repeating: 0.0, count: from.count)
        var speed = value

        for lane in value.indices {
            if velocity[lane] == 0 {
                value[lane] = from[lane] + ((destination[lane] - from[lane]) * amount)
                speed[lane] = (destination[lane] - from[lane]) * climb / length
            } else {
                value[lane] = (h00 * from[lane])
                    + (h10 * length * velocity[lane])
                    + (h01 * destination[lane])
                speed[lane] = ((d00 * from[lane])
                    + (d10 * length * velocity[lane])
                    + (d01 * destination[lane])) / length
            }
        }

        return HostMotionSample(value: value, velocity: speed, rested: false)
    }

    /// A mass on a spring, in closed form about the distance left rather than
    /// the value - which makes the three damping cases the textbook ones and
    /// keeps the destination out of the exponentials.
    private static func spring(
        response: Double,
        damping: Double,
        elapsed: Double,
        from: [Double],
        destination: [Double],
        velocity: [Double]
    ) -> HostMotionSample {
        guard elapsed < longest else { return landed(destination) }

        let frequency = 2 * Double.pi / response
        var value = Array(repeating: 0.0, count: from.count)
        var speed = value
        var rested = true

        for lane in value.indices {
            let target = destination[lane]
            let displacement = from[lane] - target
            let initialSpeed = velocity[lane]
            let remaining: Double
            let derivative: Double

            if abs(damping - 1) < 1e-6 {
                let b = initialSpeed + (frequency * displacement)
                let decay = exp(-frequency * elapsed)
                remaining = (displacement + (b * elapsed)) * decay
                derivative = (b - (frequency * (displacement + (b * elapsed)))) * decay
            } else if damping < 1 {
                let damped = frequency * sqrt(1 - (damping * damping))
                let a = displacement
                let b = (initialSpeed + (damping * frequency * displacement)) / damped
                let decay = exp(-damping * frequency * elapsed)
                let cosine = cos(damped * elapsed)
                let sine = sin(damped * elapsed)
                remaining = decay * ((a * cosine) + (b * sine))
                derivative = decay * (
                    (-damping * frequency * ((a * cosine) + (b * sine)))
                        + (damped * ((b * cosine) - (a * sine))))
            } else {
                let root = frequency * sqrt((damping * damping) - 1)
                let first = -(frequency * damping) + root
                let second = -(frequency * damping) - root
                let c1 = (initialSpeed - (second * displacement)) / (first - second)
                let c2 = displacement - c1
                remaining = (c1 * exp(first * elapsed)) + (c2 * exp(second * elapsed))
                derivative = (c1 * first * exp(first * elapsed))
                    + (c2 * second * exp(second * elapsed))
            }

            if abs(remaining) <= still && abs(derivative) <= still {
                value[lane] = target
                speed[lane] = 0
            } else {
                value[lane] = target + remaining
                speed[lane] = derivative
                rested = false
            }
        }

        return rested
            ? landed(destination)
            : HostMotionSample(value: value, velocity: speed, rested: false)
    }

    /// How far through the change an eased animation is when `progress` of its
    /// length has passed: 0 at the start and 1 at the end, and past 1 and back on the
    /// curves that overshoot.
    static func ease(_ curve: Easing, at progress: Double) -> Double {
        let value = min(max(progress, 0), 1)

        switch curve {
        case .linear:
            return value
        case .sineOut:
            return sin(value * .pi / 2)
        case .sineIn:
            return 1 - cos(value * .pi / 2)
        case .sineInOut:
            return (1 - cos(value * .pi)) / 2
        case .cubicIn:
            return value * value * value
        case .cubicOut:
            let shifted = value - 1
            return (shifted * shifted * shifted) + 1
        case .cubicInOut:
            if value < 0.5 { return 4 * value * value * value }
            let shifted = (2 * value) - 2
            return (shifted * shifted * shifted / 2) + 1
        case .bounceOut:
            return bounceOut(value)
        case .bounceIn:
            return 1 - bounceOut(1 - value)
        case .springIn:
            return value * value * ((2.70158 * value) - 1.70158)
        case .springOut:
            let shifted = value - 1
            return (shifted * shifted * ((2.70158 * shifted) + 1.70158)) + 1
        }
    }

    /// How steeply `curve` climbs at `progress`, per unit of progress: a central
    /// difference on the curve, a function of progress alone.
    static func slope(_ curve: Easing, at progress: Double) -> Double {
        let step = 1e-4
        let from = min(max(progress - step, 0), 1)
        let to = min(max(progress + step, 0), 1)
        guard to > from else { return 0 }
        return (ease(curve, at: to) - ease(curve, at: from)) / (to - from)
    }

    private static func bounceOut(_ value: Double) -> Double {
        if value < 1 / 2.75 {
            return 7.5625 * value * value
        }
        if value < 2 / 2.75 {
            let shifted = value - (1.5 / 2.75)
            return (7.5625 * shifted * shifted) + 0.75
        }
        if value < 2.5 / 2.75 {
            let shifted = value - (2.25 / 2.75)
            return (7.5625 * shifted * shifted) + 0.9375
        }

        let shifted = value - (2.625 / 2.75)
        return (7.5625 * shifted * shifted) + 0.984375
    }
}
